src/osd/ReplicatedPG.cc

```diff
@@ -4487,6 +4506,60 @@ int ReplicatedPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
       }
       break;
 
+    case CEPH_OSD_OP_CACHE_PIN:
+      tracepoint(osd, do_osd_op_pre_cache_pin, soid.oid.name.c_str(), soid.snap.val);
+      if ((!pool.info.is_tier() ||
+	  pool.info.cache_mode == pg_pool_t::CACHEMODE_NONE)) {
+        result = -EINVAL;
+        dout(10) << " pin object is only allowed on the cache tier " << dendl;
+        break;
+      }
+      ++ctx->num_write;
+      {
+        if (!obs.exists) {
+	  assert(!oi.is_cache_pinned());
+          ctx->mod_desc.create();
+          t->touch(soid);
+          ctx->delta_stats.num_objects++;
+          obs.exists = true;
+	  obs.oi.new_object();
```

Comment:
We can change this to ENOENT.
PIN/UNPIN are write ops. I remember once there is a mail from you in the mail list, saying that a write op implicitly creates the object if it doesn't exist?